Mixed Content

kristijanburnik opened the pull request:
Initial test specification, basic assertions and generating tools

@mikewest

kristijanburnik commented:
Sry, I again wrote @mkwst instead of @mikewest. :-/

PR is a bit sketchy at this point with rough edges all over the place, 
but I would like some of your feedback nevertheless.

kristijanburnik committed:
Mixed-Content: Add initial mixed-content tests and maintenance tools.

kristijanburnik commented:
Some cleanup from previous commit. 
Implemented resource querying. 
This PR should now generate 780 tests, which will be reduced.

@mikewest

mikewest commented:
What patches do I need to apply locally to get the tests to run?

kristijanburnik commented:
Assuming in ~/web-platform-tests, apply these patches:

in ./tools:
https://github.com/w3c/wpt-tools/pull/15.patch

in ./tools/wptserve
https://github.com/w3c/wptserve/pull/59.patch

in ./
https://github.com/w3c/web-platform-tests/pull/1889.patch


And then generate the tests:

cd ./mixed-content
./generic/tools/generate.py
git add * && git commit -m "IGF"
../manifest

Spin them:

google-chrome https://web-platform.test:8443/tools/runner/index.html
|| chrome https://web-platform.test:8443/tools/runner/index.html

Enter directory:
/mixed-content

Run the tests.





On Mon, Jun 22, 2015 at 2:32 PM, Mike West <notifications@github.com> wrote:

> What patches do I need to apply locally to get the tests to run?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/1889#issuecomment-114086348>
> .
>



-- 

*Kristijan Burnik*

Software Engineering Intern

burnik@google.com

Google Germany GmbH

Dienerstraße 12

80331 München

Geschäftsführer: Graham Law, Christine Elizabeth Flores

Registergericht und -nummer: Hamburg, HRB 86891

Sitz der Gesellschaft: Hamburg

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+function normalizePort(targetPort) {
+  var defaultPorts = [80, 443];
+  var isDefaultPortForProtocol = (defaultPorts.indexOf(targetPort) >= 0);
```
This tells you if a port is a default port, but not if it's the right default port. That is, `http://example.com:443` would normalize to "" with this code, which is incorrect.

kristijanburnik replied:
I would say the problem here is the name of the variable ```isDefaultPortForProtocol```. It should be ```isDefaultPort``` and that would make things more clear.

So normalizePort doesn't care about the protocol it just removes the :80 or :443 if they are target ports.

I'm mapping the corresponding protocol in a different place. See in mixed-content-test-case.js:

```
var insecurePort = normalizePort(parseInt("{{ports[http][0]}}"));
  var securePort = normalizePort(parseInt("{{ports[https][0]}}"));

```

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+function setAttributes(el, attrs) {
+  attrs = attrs || {}
+  for (var attr in attrs)
+    el[attr] = attrs[attr];
```
Assuming `el` is an Element, I'd prefer `el.setAttribute(attr, attrs[attr])`.

kristijanburnik replied:
Done.

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+function createElement(tagName, attrs, parent) {
+  var el = document.createElement(tagName);
+  setAttributes(el, attrs);
+  if (parent && parent.appendChild)
```
When would this come into play?

kristijanburnik replied:
Well, every time I specify a parent for the element that gets created. For example a source tag as child of a picture tag.

mikewest replied:
I meant the other bit: When would that parent tag not have an `appendChild` method?

kristijanburnik replied:
Haha, right... residue from a previous experiment which is not in play any more.
Removed.

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+  return createElement("iframe", {"name": name}, document.body);
+}
+
+function queryIframe(url) {
```
Nit: Perhaps `createQueryIFrame`? Without `create`, it sounds like you're querying something, so I'd expect `querySelector`. :)

mikewest replied:
Nevermind. It wasn't clear to me that `query*` is what you're using to denote tests. Perhaps `testX` instead?

kristijanburnik replied:
What about invokeFoo or loadViaFoo or requestViaFoo ?

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+}
+
+function queryImage(url) {
+  createElement("img", {"src": url}, document.body)
```
Ditto.

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+  var xhr = new XMLHttpRequest();
+  xhr.open('GET', url, true);
+  xhr.responseType = "json";
+  xhr.send();
```
It surprises me that this is enough. Shouldn't you need to bind to various events to read the XHR state?

kristijanburnik replied:
If we look at all types of resources we cover, I think there isn't really a nice way to get callbacks for each (e.g. did the request go through or not). E.g. if we load an image via CSS.

 I'll see how much we can cover and possibly add additional checks. In any case, some onload events and other XHR events should be least of the problems.

The test itself is reusing server-side state to confirm if the requests went through or not. That's the overall idea, however the timeout for the assertion after the request can be a potential cause of flakiness.

mikewest replied:
Understood. However, for developers it's important to get cross-browser agreement about what the web-visible side-effects of mixed content blocking are for various resource types. That is, browser should behave as though there was a network error, which has certain impacts on events, etc. The tests should reflect that, regardless of the reality between browsers. Then we fix the browsers. :)

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+function queryWorker(url) {
+  var worker = new Worker(url);
+  console.log("worker", url);
+  worker.postMessage('');
```
Ditto.

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+    fetch(url);
+  } catch (ex) {
+    console.error("Failed: fetch constructor.", ex);
+  }
```
Ditto.

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+  try {
+    fetch(url);
+  } catch (ex) {
+    console.error("Failed: fetch constructor.", ex);
```
Catching here with a console error seems like a remnant of debugging?

kristijanburnik replied:
I believe I put this here so that the promises don't fail due to the exception thrown in the test.
Added a TODO to check it out.

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+           .toString(16).substring(1);
+  }
+  return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() +
+         s4() + s4();
```
This looks like it was taken from http://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript. The second answer there seems better.

mikewest replied:
(And please note the source in a comment).

kristijanburnik replied:
Done.

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+function normalizePort(targetPort) {
```
Sorry that I didn't ask you to add comments to the `referrer-policy` code. I didn't do a great job reviewing... I'll pay more attention this time. :)

Please add comments to this code. I'd suggest following the guidelines at http://google.github.io/styleguide/javascriptguide.xml#Comments.

kristijanburnik replied:
Added a jsDoc for this method.

mikewest replied:
(Just so it's clear, by "this code", I meant "all the code in this PR") :)

kristijanburnik replied:
I'll add more comments soon. :)

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+        resolve({roundtrip_ms:roundtrip_ms, index:index});
+      };
+
+      startTime = (new Date()).getTime();
```
I'd suggest folding this into line 162.

kristijanburnik replied:
Done.

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+  function getStats(results) {
+    var stats = {
+      sum: 0,
+      min: 100000,
```
Nit: `Number.MAX_VALUE`.

kristijanburnik replied:
Done.

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+    var stats = {
+      sum: 0,
+      min: 100000,
+      max: -100000,
```
Nit: `Number.MIN_VALUE`.

kristijanburnik replied:
Done.

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,238 @@
+    for (var i = 0; i < numMeasurements - 1; ++i) {
+      p = p.then(function(result) {
+        results.push(result.roundtrip_ms);
+        return measureRoundtripTime(url, result.index + 1);
```
This ends up being a _lot_ of extra requests over the course of several hundred tests. I worry about what it's going to do to both test stability and overall runtime.

kristijanburnik replied:
Yes, I am still thinking about this approach. I wouldn't however be very satisfied by setting some fixed timeout for the resource... 

I would consider not doing this measurement on every test, however it's more likely the subsequent requests would take approx the same time if they're closer together, I'd say. For example, if the machine running the tests for some (external) reason gets a higher load on a test at index 100 and a response time takes 3x longer than the first 100 tests, the affected test should most likely still work fine as we're taking the response ETA at that busy time chunk.

Currently I only make 3 extra requests per test, which is not that horrible. They are also done in sequence.

kristijanburnik replied:
It might happen that most of the resources will not require these heuristic timeouts. I'm still playing around with that.

kristijanburnik commented:
Refactored the test suite:
* Reused onload & onerror events instead of timeoutPromise where applicable
* Fixed tests with media
* renamed methods

@mikewest

mikewest commented on mixed-content/generic/mixed-content-test-case.js:
```diff
@@ -0,0 +1,224 @@
+function MixedContentTestCase(scenario, description, sanityChecker) {
+  var insecureProtocol = "http:"
```
Tiny nit: The protocol is "http". The `:` separates the protocol from the rest of the URL. :)

kristijanburnik replied:
I was using this format so it's comparable to location.protocol - comparing is easier. Any idea why location.protocol is using a colon?
 
Refactored.

mikewest replied:
/me waves at @annevk: Your API makes no sense! :)

mikewest commented on mixed-content/generic/mixed-content-test-case.js:
```diff
@@ -0,0 +1,224 @@
+    "same-origin":
+      location.origin + resourcePath,
+    "same-host-https":
+      secureProtocol + "//" + sameOriginHost + securePort + resourcePath,
```
You need a `:` between the host and port, right? (e.g. `example.com_:_123`)

kristijanburnik replied:
Yes, `getNormalizedPort` prepends a colon. If a default or no port is used, it evaluates to an empty string. Added a JsDoc to the func. 

Scope of securePort and insecurePort is fairly small, however I am not sure how to name it properly: E.g. secureMaybePort, securePortish, securePortClause, securePortPortion all sound pretty bad. I suppose I'll just add a comment what the values can be.

mikewest commented on mixed-content/generic/mixed-content-test-case.js:
```diff
@@ -0,0 +1,224 @@
+  var sameOriginHost = location.hostname;
+  var crossOriginHost = "{{domains[www1]}}";
+
+  var insecurePort = getNormalizedPort(parseInt("{{ports[http][0]}}"));
```
Nit: `parseInt(..., 10)`

kristijanburnik replied:
Done.

mikewest commented on mixed-content/generic/mixed-content-test-case.js:
```diff
@@ -0,0 +1,224 @@
+      insecureProtocol + "//" + crossOriginHost + insecurePort + resourcePath
+  };
+
+  var resourceInvoker = {
```
Nit: "Invoker" is a bit strange. "Fetcher", perhaps?

kristijanburnik replied:
Renamed to resourceMap since the invoking method is now called `requestResource`:

```
  var requestResource = resourceMap[scenario.subresource];
```

mikewest commented on mixed-content/generic/mixed-content-test-case.js:
```diff
@@ -0,0 +1,224 @@
+    "link-prefetch-tag": requestViaLinkPrefetch
+  };
+
+  sanityChecker.checkScenario(scenario, resourceInvoker);
```
Is this conditional on the `debug` state?

kristijanburnik replied:
Does nothing in release mode:

See common.js
```
// SanityChecker does nothing in release mode. See sanity-checker.js for debug
// mode.
function SanityChecker() {}
SanityChecker.prototype.checkScenario = function() {};

```

mikewest commented on mixed-content/generic/mixed-content-test-case.js:
```diff
@@ -0,0 +1,224 @@
+        if ("eventPromise" in queriedObject)
+          return queriedObject.eventPromise;
+        else
+          return timeoutPromise(estimatedResponseTime);
```
This timeout business continues to worry me in terms of time spent in tests. It's worth trying out to see what the impact is, but I suspect we're going to have to cull the tests down to use it, especially if CSP does the same.

kristijanburnik replied:
I'm seriously narrowing down the list of resources that will actually rely on the timeouts.
The impact shouldn't be too large overall. The estimated average for the response time is <50 ms on my machine.

mikewest commented on mixed-content/generic/common.js:
```diff
@@ -0,0 +1,268 @@
+    el.setAttribute(attr, attrs[attr]);
+}
+
+function createElement(tagName, attrs, parent, bindEvents) {
```
Passing booleans is tough to read; I'd suggest turning this into an enum-style object thing (e.g. https://developers.google.com/closure/compiler/docs/js-for-compiler#tag-enum).

kristijanburnik commented:
Added comments to common.js
Removed the need for timeouts - Only link prefetch is failing because of it. 

Total tests: 778

Update on failing tests:
CR: 60 / 778 - explained [here](https://github.com/kristijanburnik/web-platform-tests/blob/mixed-content/mixed-content/known-chrome-failures.txt)
FF: 87 / 778 - still looking at issues

@mikewest

mikewest commented on tools:
```diff
@@ -1 +1 @@
-Subproject commit 72c8068067c68dcaa0768c1d1f394738b4e60eb9
+Subproject commit 542410209352812448c5deb0bb5801cbd4a13b8a
```
The CI is failing on this.

mikewest commented on mixed-content/known-chrome-failures.txt:
```diff
@@ -0,0 +1,65 @@
+# IFRAME doesn't trigger events properly in Chrome.
```
Why include this file? I agree that it's interesting, but it's better done as bugs filed against Chrome rather than a file in the W3C repository.

kristijanburnik replied:
Removing it before I squash. I  wanted you to take a look before we land.

mikewest commented on mixed-content/generic/worker.js:
```diff
@@ -0,0 +1,7 @@
+  postMessage('done');
+} catch(err) {
+  // Works cross-origin.
+  postMessage('done', '*');
```
Why do the `try/catch` dance rather than just posting with `*`?

kristijanburnik replied:
When on same origin: ```postMessage('done', '*');``` produces this interesting error:

```Uncaught TypeError: Failed to execute 'postMessage' on 'WorkerGlobalScope': The 2nd argument is neither an array, nor does it have indexed properties.```

Bug, no bug? Haven't yet tried on FF.

mikewest replied:
Sorry, I wasn't thinking: how do you ever have a cross-origin case here? Dedicated (and shared) workers are same-origin only, aren't they?

kristijanburnik replied:
Cross origin is HTTPS --> HTTP.

We can examine it further, if you like, I can extract out just those tests to show what the issue is, at least in Chrome. Might be worthwhile investigating this further. :-)

mikewest replied:
1. I don't believe it's possible to load an HTTPS worker on an HTTP page (see step #5 of https://html.spec.whatwg.org/multipage/workers.html#dom-worker (note that Firefox allows `data:` Workers, but it also treats `data:` as the same-origin as its embedder; Chrome does not, and does not)).

2. Regardless, a worker's `postMessage` always talks to its embedder; it doesn't have an origin flag (the second argument is a set of objects to transfer from one context to the other: see https://html.spec.whatwg.org/multipage/workers.html#dedicated-workers-and-the-dedicatedworkerglobalscope-interface:dom-dedicatedworkerglobalscope-postmessage).

You can simplify this file to `postMessage('done');`.

kristijanburnik replied:
Done!

So to conclude, the issue was:

I'm mapping the content type (mime type) to what expect.py will end up
producing, since application/javascript mapped to worker.js, it was being
executed in the top-level context and postMessage was not happy about it.

Therefore even an empty catch block would have worked fine.

So for script-tag I'm now just using text/plain since we catch onload and
onerror events anyway. In any other case I would have to add another
parameter other than contentType.



On Wed, Jul 1, 2015 at 12:08 PM, Mike West <notifications@github.com> wrote:

> In mixed-content/generic/worker.js
> <https://github.com/w3c/web-platform-tests/pull/1889#discussion_r33665249>
> :
>
> > @@ -0,0 +1,7 @@
> > +try {
> > +  // Works on same origin.
> > +  postMessage('done');
> > +} catch(err) {
> > +  // Works cross-origin.
> > +  postMessage('done', '*');
>
>
>    1.
>
>    I don't believe it's possible to load an HTTPS worker on an HTTP page
>    (see step #5 <https://github.com/w3c/web-platform-tests/pull/5> of
>    https://html.spec.whatwg.org/multipage/workers.html#dom-worker (note
>    that Firefox allows data: Workers, but it also treats data: as the
>    same-origin as its embedder; Chrome does not, and does not)).
>    2.
>
>    Regardless, a worker's postMessage always talks to its embedder; it
>    doesn't have an origin flag (the second argument is a set of objects to
>    transfer from one context to the other: see
>    https://html.spec.whatwg.org/multipage/workers.html#dedicated-workers-and-the-dedicatedworkerglobalscope-interface:dom-dedicatedworkerglobalscope-postmessage
>    ).
>
> You can simplify this file to postMessage('done');.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/1889/files#r33665249>.
>



-- 

*Kristijan Burnik*

Software Engineering Intern

burnik@google.com

Google Germany GmbH

Dienerstraße 12

80331 München

Geschäftsführer: Graham Law, Christine Elizabeth Flores

Registergericht und -nummer: Hamburg, HRB 86891

Sitz der Gesellschaft: Hamburg

mikewest commented on mixed-content/generic/expect.py:
```diff
@@ -0,0 +1,99 @@
+    scheme = parsed.scheme
+
+    if swap_scheme:
+      if parsed.scheme == "http":
```
What about HTTPS -> HTTP?

This could be done as the following, for example:

```
if swap_scheme:
  scheme = "http" if parsed.scheme == "https" else "https"
```

kristijanburnik replied:
Haha, I've actually had the same code and for some reason decided to refactor to the current state. Agreed! Putting it back. :)

mikewest commented on mixed-content/generic/mixed-content-test-case.js:
```diff
@@ -0,0 +1,169 @@
+    "form-tag": "text/html",
+    "iframe-tag": "text/html",
+    "img-tag":  "image/png",
+    "script-tag": "text/plain",
```
Can you pick an executable-as-javascript MIME type? That is, one of these: https://code.google.com/p/chromium/codesearch#chromium/src/components/mime_util/mime_util.cc&rcl=1435735758&l=59

mikewest commented:
LGTM % the MIME type and the Travis CI error.

Do the submodule updates depend on changes that haven't landed yet?

kristijanburnik commented:
Ok, changing to text/javascript for scripts.
Leaving application/javascript for workers.

In short, yes. I depend on commits on my branches which are in review.
The merge commit will differ so we cannot land before we handle shared
stash to upstream master. I can then rewind the heads to the merge commit
positions and repush.

If we were to land now, I could check if there is shared_stash and use the
regular stash if not there, but that will just fail a bunch of tests until
@jgraham and I figure it out with the new shiny SharedStash feature. :-)



On Wed, Jul 1, 2015 at 1:32 PM, Mike West <notifications@github.com> wrote:

> LGTM % the MIME type and the Travis CI error.
>
> Do the submodule updates depend on changes that haven't landed yet?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/1889#issuecomment-117620504>
> .
>



-- 

*Kristijan Burnik*

Software Engineering Intern

burnik@google.com

Google Germany GmbH

Dienerstraße 12

80331 München

Geschäftsführer: Graham Law, Christine Elizabeth Flores

Registergericht und -nummer: Hamburg, HRB 86891

Sitz der Gesellschaft: Hamburg

kristijanburnik commented:
While we wait on the SharedStash, I've refactored the request API to return promises rather than the elements. I also fixed a bug with form submission (bad event binding).

@mikewest

mikewest commented:
Yay. Let's add a few minutes to the cycle time of various test bots and see what explodes.

/cc @jgraham as a heads up that a few hundred new tests will be hitting Mozilla's test bots shortly, and that the form tests in particular are probably going to cause some problems given Firefox's form submission modal dialog.

kristijanburnik commented:
These are the tests which do the form submission.

/mixed-content/allowed/no-opt-in/same-host-https/form-tag/top-level/keep-scheme-redirect/allowed.https.html
/mixed-content/allowed/no-opt-in/same-host-https/form-tag/top-level/no-redirect/allowed.https.html
/mixed-content/allowed/http-csp/same-host-https/form-tag/top-level/keep-scheme-redirect/allowed.https.html
/mixed-content/allowed/http-csp/same-host-https/form-tag/top-level/no-redirect/allowed.https.html
/mixed-content/allowed/meta-csp/same-host-https/form-tag/top-level/no-redirect/allowed.https.html
/mixed-content/blockable/no-opt-in/same-host-http/form-tag/top-level/keep-scheme-redirect/no-opt-in-blocks.https.html
/mixed-content/blockable/no-opt-in/same-host-http/form-tag/top-level/no-redirect/no-opt-in-blocks.https.html
/mixed-content/blockable/no-opt-in/same-host-http/form-tag/top-level/swap-scheme-redirect/no-opt-in-blocks.https.html
/mixed-content/blockable/no-opt-in/cross-origin-http/form-tag/top-level/keep-scheme-redirect/no-opt-in-blocks.https.html
/mixed-content/blockable/no-opt-in/cross-origin-http/form-tag/top-level/no-redirect/no-opt-in-blocks.https.html
/mixed-content/blockable/no-opt-in/cross-origin-http/form-tag/top-level/swap-scheme-redirect/no-opt-in-blocks.https.html
/mixed-content/blockable/http-csp/same-host-http/form-tag/top-level/keep-scheme-redirect/opt-in-blocks.https.html
/mixed-content/blockable/http-csp/same-host-http/form-tag/top-level/no-redirect/opt-in-blocks.https.html
/mixed-content/blockable/http-csp/same-host-http/form-tag/top-level/swap-scheme-redirect/opt-in-blocks.https.html
/mixed-content/blockable/http-csp/cross-origin-http/form-tag/top-level/keep-scheme-redirect/opt-in-blocks.https.html
/mixed-content/blockable/http-csp/cross-origin-http/form-tag/top-level/no-redirect/opt-in-blocks.https.html
/mixed-content/blockable/http-csp/cross-origin-http/form-tag/top-level/swap-scheme-redirect/opt-in-blocks.https.html
/mixed-content/blockable/meta-csp/same-host-http/form-tag/top-level/no-redirect/opt-in-blocks.https.html
/mixed-content/blockable/meta-csp/cross-origin-http/form-tag/top-level/no-redirect/opt-in-blocks.https.html

mikewest commented:
Thanks, @kristijanburnik.

/cc @dveditz @tanvihacks @annevk
